### PR TITLE
[FLINK-28202][ci] Generalize utils around shade plugin 

### PR DESCRIPTION
**Pull request by zentol**
~~Based on #20067.~~

Similar to #20053, just for the shade plugin parsing logic.

### CODE REVIEW

File: tools/ci/java-ci-tools/src/main/java/org/apache/flink/tools/ci/utils/shade/ShadeParser.java [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+import java.util.Set;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+import java.util.stream.Stream;
+
+/** Utils for parsing the shade-plugin output. */
+public final class ShadeParser {
+
+    private static final Pattern SHADE_NEXT_MODULE_PATTERN =
+            Pattern.compile(".*:shade \\((?:shade-flink|shade-dist)\\) @ ([^ _]+)(_[0-9.]+)? --.*");
```
Comment (zentol) [inline tools/ci/java-ci-tools/src/main/java/org/apache/flink/tools/ci/utils/shade/ShadeParser.java:40]:
```suggestion
            Pattern.compile(".*:shade \\((?:shade-flink|shade-dist|default)\\) @ ([^ _]+)(_[0-9.]+)? --.*");
```

**Review by alpinegizmo — APPROVED** [review]
LGTM +1

**Comment by alpinegizmo** [comment]
I see that the build failed on the spotless check.

**Review by alpinegizmo — COMMENTED** [review]
Code looks fine to me, but spotless seems to be unhappy.

**Review by alpinegizmo — CHANGES_REQUESTED** [review]
If I read the build output correctly, spotless didn't like something.

**Comment by zentol** [comment]
Still a gremlin in there somewhere in the parsing...

File: tools/ci/flink-ci-tools/src/main/java/org/apache/flink/tools/ci/utils/shade/ShadeParser.java [hunk 2]
```diff
@@ -46,7 +46,6 @@ public final class ShadeParser {
                             + "(?<groupId>.*?):"
                             + "(?<artifactId>.*?):"
                             + "(?<type>.*?):"
-                            + "(?:(?<classifier>.*?):)?"
```
Comment (zentol) [inline tools/ci/flink-ci-tools/src/main/java/org/apache/flink/tools/ci/utils/shade/ShadeParser.java]: Had to roll-back this addition to keep the behavior in sync with the NoticeParser, which currently doesn't account for classifiers and just parses them as part of the version.

Will be fixed in a follow-up.

File: tools/ci/flink-ci-tools/src/test/java/org/apache/flink/tools/ci/utils/dependency/DependencyParserTreeTest.java [hunk 3]
```diff
@@ -37,7 +37,7 @@ private static Stream<String> getTestDependencyTree() {
                 "[INFO] +- external:dependency1:jar:2.1:compile",
                 "[INFO] |  +- external:dependency2:jar:2.2:compile (optional)",
                 "[INFO] |  |  \\- external:dependency3:jar:2.3:provided",
-                "[INFO] |  +- external:dependency4:classifier:jar:2.4:compile",
+                "[INFO] |  +- external:dependency4:jar:classifier:2.4:compile",
```
Comment (zentol) [inline tools/ci/flink-ci-tools/src/test/java/org/apache/flink/tools/ci/utils/dependency/DependencyParserTreeTest.java:40]: The order was actually wrong :see_no_evil:

File: tools/ci/flink-ci-tools/src/main/java/org/apache/flink/tools/ci/utils/shared/Dependency.java [hunk 4]
```diff
@@ -121,12 +122,13 @@ public boolean equals(Object o) {
         return Objects.equals(groupId, that.groupId)
                 && Objects.equals(artifactId, that.artifactId)
                 && Objects.equals(version, that.version)
+                && Objects.equals(classifier, that.classifier)
```
Comment (zentol) [inline tools/ci/flink-ci-tools/src/main/java/org/apache/flink/tools/ci/utils/shared/Dependency.java:125]: this hid some test failures :see_no_evil:

File: tools/ci/flink-ci-tools/src/main/java/org/apache/flink/tools/ci/utils/shade/ShadeParser.java (hunk 2 above)
Comment (alpinegizmo) [reply tools/ci/flink-ci-tools/src/main/java/org/apache/flink/tools/ci/utils/shade/ShadeParser.java]: Sure, okay. Baby steps.

**Pull request merged**